packages/firestore/test/unit/local/indexeddb_schema.test.ts

```diff
@@ -0,0 +1,153 @@
+} from '../../../src/local/indexeddb_schema';
+import { SimpleDb, SimpleDbTransaction } from '../../../src/local/simple_db';
+import { PersistencePromise } from '../../../src/local/persistence_promise';
+
+const INDEXEDDB_TEST_DATABASE = 'schemaTest';
+
+function withDb(
+  schemaVersion,
+  fn: (db: IDBDatabase) => Promise<void>
+): Promise<void> {
+  return new Promise<IDBDatabase>((resolve, reject) => {
+    const request = window.indexedDB.open(
+      INDEXEDDB_TEST_DATABASE,
+      schemaVersion
+    );
+    request.onupgradeneeded = (event: IDBVersionChangeEvent) => {
+      const db = (event.target as IDBOpenDBRequest).result;
+      createOrUpgradeDb(
+        db,
+        new SimpleDbTransaction(request.transaction),
+        event.oldVersion,
+        schemaVersion
+      );
+    };
+    request.onsuccess = (event: Event) => {
+      resolve((event.target as IDBOpenDBRequest).result);
+    };
+    request.onerror = (event: ErrorEvent) => {
+      reject((event.target as IDBOpenDBRequest).error);
+    };
+  })
+    .then(db => fn(db).then(() => db))
+    .then(db => {
+      db.close();
+    });
+}
+
+function getAllObjectStores(db: IDBDatabase): String[] {
+  const objectStores: String[] = [];
+  for (let i = 0; i < db.objectStoreNames.length; ++i) {
+    objectStores.push(db.objectStoreNames.item(i));
+  }
+  objectStores.sort();
+  return objectStores;
+}
+
+function getTargetCount(db: IDBDatabase): Promise<number> {
+  const sdb = new SimpleDb(db);
+  return sdb
+    .runTransaction('readonly', [DbTargetGlobal.store], txn =>
+      txn
+        .store<DbTargetGlobalKey, DbTargetGlobal>(DbTargetGlobal.store)
+        .get(DbTargetGlobal.key)
+    )
+    .then(metadata => metadata.targetCount);
+}
+
+describe('IndexedDbSchema: createOrUpgradeDb', () => {
+  if (!IndexedDbPersistence.isAvailable()) {
+    console.warn('No IndexedDB. Skipping createOrUpgradeDb() tests.');
+    return;
+  }
+
+  beforeEach(() => SimpleDb.delete(INDEXEDDB_TEST_DATABASE));
+
+  it('can install schema version 1', () => {
+    return withDb(1, async db => {
+      expect(db.version).to.equal(1);
+      // Version 1 adds all of the stores so far.
+      expect(getAllObjectStores(db)).to.have.members(V1_STORES);
+    });
+  });
+
+  it('can install schema version 2', () => {
+    return withDb(2, db => {
+      expect(db.version).to.equal(2);
+      // We should have all of the stores, we should have the target global row
+      // and we should not have any targets counted, because there are none.
+      expect(getAllObjectStores(db)).to.have.members(V2_STORES);
+      // Check the target count. We haven't added any targets, so we expect 0.
+      return getTargetCount(db).then(targetCount => {
+        expect(targetCount).to.equal(0);
+      });
+    });
+  });
+
+  it('can install schema version 3', () => {
+    return withDb(3, async db => {
+      expect(db.version).to.be.equal(3);
+      expect(getAllObjectStores(db)).to.have.members(ALL_STORES);
```

Comment:
Are you intentionally using ALL_STORES instead of V3_STORES?  I'd be inclined to use V3_STORES so these tests don't need to change over time as we add new stores in later schema versions...

Reply:
I was thinking of changing this already, but kept it the same as before. Fixed now.